### PR TITLE
Reified generic and stuff

### LINKED ISSUE #2: add reified generics

### PR DESCRIPTION
**Pull request by DetachHead**
closes #2

### CODE REVIEW

**Comment by KotlinIsland** [comment]
Woow, all this stuff looks based

File: .pylintrc [hunk 1]
```diff
@@ -65,22 +67,15 @@ confidence=
...
+      missing-module-docstring,
+      missing-class-docstring,
+      invalid-name, # https://github.com/PyCQA/pylint/issues/3401
+      fixme,
+      useless-import-alias, # required when using no-implicit-reexport in mypy
+      no-member, # handled by mypy
+      inherit-non-class # handled by mypy, false positive on ReifiedGeneric
 
 # Enable the message, report, category or checker with the given id(s). You can
 # either give multiple identifier separated by comma (,) or put this option
```
Comment (KotlinIsland) [inline .pylintrc]: Doesn't pylint support pyproject.toml?

File: basedtyping/callables.py [hunk 2]
```diff
@@ -0,0 +1,3 @@
+from typing import Callable
+
+AnyFunction = Callable[..., object]  # type:ignore[misc]
```
Comment (KotlinIsland) [inline basedtyping/callables.py]: Why do you think `AnyFunction` is a good name?

Also could have some docs on it explaining why this hack is needed.

File: basedtyping/generics.py [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+
+from typing import TypeVar
+
+from basedtyping.callables import AnyFunction
+
+T = TypeVar("T")
+T_co = TypeVar("T_co", covariant=True)
+T_cont = TypeVar("T_cont", contravariant=True)
+Fn = TypeVar("Fn", bound=AnyFunction)
+Self = TypeVar("Self")
```
Comment (KotlinIsland) [inline basedtyping/generics.py]: Does this actually work? I though you needed to bind each usage to the current class?

**Comment by DetachHead** [comment]
should be mostly ready now, just gotta do a couple more things
- [x] add documentation to the `ReifiedGeneric` class
- [x] add tests for variance (like the ones i wrote in my original impl)

File: tests/test_runtime_checks.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
+from basedtyping.runtime_checks import is_subclass
+
+# pylint:disable=no-self-use
+
+
+class TestIsSubclass:
```
Comment (KotlinIsland) [inline tests/test_runtime_checks.py]: Why is this a class

File: basedtyping/callables.py (hunk 2 above)
Comment (DetachHead) [reply basedtyping/callables.py]: What’s wrong with the name, what would u suggest

File: basedtyping/generics.py (hunk 3 above)
Comment (DetachHead) [reply basedtyping/generics.py]: I don’t remember what your edge case was that required doing it that way, but in most cases this works just fine

File: tests/test_runtime_checks.py (hunk 4 above)
Comment (DetachHead) [reply tests/test_runtime_checks.py]: Just to group the tests

Comment (DetachHead) [reply tests/test_runtime_checks.py]: there was another function in `runtime_checks` so it was 2 classes, but i removed it. it looked like this
image

File: basedtyping/callables.py (hunk 2 above)
Comment (DetachHead) [reply basedtyping/callables.py]: renamed to `Function` and added documentation

File: basedtyping/generics.py (hunk 3 above)
Comment (KotlinIsland) [reply basedtyping/generics.py]: https://mypy-play.net/?mypy=latest&python=3.10&flags=strict&gist=9ae0ae81e13d19f43eb46948d7041747

Comment (KotlinIsland) [reply basedtyping/generics.py]: I would prefer not to expose a half baked thing that will be changed for certain later.

File: basedtyping/runtime_checks.py [hunk 5]
```diff
@@ -0,0 +1,13 @@
+from types import UnionType
+from typing import Sequence, cast
+
+
+def is_subclass(type1: type | UnionType, type2: type | UnionType) -> bool:
```
Comment (KotlinIsland) [inline basedtyping/runtime_checks.py]: I wonder if `issubform` would be a better name? because this is comparing unions which aren't classes/`type`s.

Regardless, needs a better doc to specify what it does.

File: basedtyping/internal_typing_stubs.py [hunk 6]
```diff
@@ -0,0 +1,90 @@
+"""stubs for the internal "type machinery" defined in the typing module.
```
Comment (KotlinIsland) [inline basedtyping/internal_typing_stubs.py]: What's the purpose of this?

File: basedtyping/type_checks.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+from basedtyping.generics import T
+
+
+class assert_type(Generic[T]):
+    """used to assert at type-time that a value is type ``T``
+
+    this is more like a function than a class, but it's defined as a class so that you can explicitly specify the generic"""
+
+    # None return type on __new__ is supported in pyright but not mypy
+    def __new__(cls, value: T) -> None:  # type:ignore[misc]
```
Comment (KotlinIsland) [inline basedtyping/type_checks.py]: Why not init?

File: basedtyping/type_checks.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+from __future__ import annotations
+
+from typing import Generic
+
+from basedtyping.generics import T
+
+
+class assert_type(Generic[T]):
```
Comment (KotlinIsland) [inline basedtyping/type_checks.py]: Is this for use in type code and tests? Maybe doc its use cases.

File: basedtyping/generics.py (hunk 3 above)
Comment (DetachHead) [reply basedtyping/generics.py]:
```py
from typing import TypeVar

Self = TypeVar("Self")

class A:
    def foo(self: Self) -> Self:
        return self
class B:
    def foo(self: Self) -> Self:
        return self     
```
this works

File: basedtyping/runtime_checks.py (hunk 5 above)
Comment (DetachHead) [reply basedtyping/runtime_checks.py]: `issubtype`? also whhats ur opinion on `is_subtype` vs `issubtype`, nothing in this language is consistent so what do we do?

File: tests/test_callables.py [hunk 9]
```diff
@@ -0,0 +1,9 @@
+from typing import TYPE_CHECKING
+
+from basedtyping.callables import Function
+from basedtyping.type_checks import assert_type
+
+if TYPE_CHECKING:
```
Comment (KotlinIsland) [inline tests/test_callables.py]: Why is this defined in type checking?

Oh, these are only tested within mypy, not pytest?

Might be good to document that.

File: tests/test_callables.py [hunk 10]
```diff
@@ -0,0 +1,9 @@
...
+
+    def test_function_type() -> None:
+        assert_type[Function](lambda: ...)
```
Comment (KotlinIsland) [inline tests/test_callables.py]: Would be good to also check other types of functions such as `builtin_function` and the various wrappers. I think they are all in `types`.

File: tests/test_reified_generic.py [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+        assert isinstance(Foo[int | str](), Foo[int])  # type:ignore[misc]
+        assert not isinstance(Foo[int](), Foo[int | str])  # type:ignore[misc]
+
+    def test_invariant(self) -> None:
+        class Foo(ReifiedGeneric[T]):
+            pass
+
+        assert not isinstance(Foo[int](), Foo[int | str])  # type:ignore[misc]
+        assert not isinstance(Foo[int | str](), Foo[int])  # type:ignore[misc]
```
Comment (KotlinIsland) [inline tests/test_reified_generic.py]: Some tests with typevars as the parameters would be good

File: basedtyping/runtime_checks.py [hunk 12]
```diff
@@ -0,0 +1,20 @@
+from types import UnionType
+from typing import Sequence, cast
+
+
+def issubtype(type1: type | UnionType, type2: type | UnionType) -> bool:
```
Comment (KotlinIsland) [inline basedtyping/runtime_checks.py]: `issubform`?

File: basedtyping/runtime_checks.py (hunk 5 above)
Comment (KotlinIsland) [reply basedtyping/runtime_checks.py]: `issubform` is a good one imo. kingsized form of `form` is `typeform` means as 'type time type' as opposed to `type` meaning as 'runtime class'

File: tests/test_callables.py (hunk 9 above)
Comment (KotlinIsland) [reply tests/test_callables.py]: Down the line would be better to make it an explicit pytest instead of just a mypy check.

File: basedtyping/internal_typing_stubs.py (hunk 6 above)
Comment (DetachHead) [reply basedtyping/internal_typing_stubs.py]: so that `_GenericAlias` is typed. alternative is suppressing a ton of mypy errors in `reified_generic.py`, i figured it's better to just type these internal things so they're easier to use

File: basedtyping/type_checks.py (hunk 7 above)
Comment (DetachHead) [reply basedtyping/type_checks.py]: because it's a hacky fake class thats more like a function, where it doesnt actually return an instance.

it won't be needed once https://github.com/KotlinIsland/basedmypy/issues/121 or https://github.com/KotlinIsland/basedmypy/issues/122 is done

File: tests/test_callables.py (hunk 9 above)
Comment (DetachHead) [reply tests/test_callables.py]: > Down the line would be better to make it an explicit pytest instead of just a mypy check.

how would i do that?

File: basedtyping/runtime_checks.py (hunk 12 above)
Comment (DetachHead) [reply basedtyping/runtime_checks.py]: resolving this because we already have [this thread](https://github.com/KotlinIsland/basedtyping/pull/7#pullrequestreview-842378852)

**Comment by DetachHead** [comment]
need to move `internal_typing_stubs` to basedtypeshed but can't because idk how to use basedtypeshed in this project. https://github.com/KotlinIsland/basedtypeshed/pull/3

File: basedtyping/generics.py (hunk 3 above)
Comment (DetachHead) [reply basedtyping/generics.py]: removed since basedmypy will have a better one

File: basedtyping/internal_typing_stubs.py (hunk 6 above)
Comment (DetachHead) [reply basedtyping/internal_typing_stubs.py]: need to move `internal_typing_stubs` to basedtypeshed but can't because idk how to use basedtypeshed in this project. https://github.com/KotlinIsland/basedtypeshed/pull/3

File: basedtyping/runtime_checks.py (hunk 5 above)
Comment (DetachHead) [reply basedtyping/runtime_checks.py]: renamed to `issubform`

File: basedtyping/internal_typing_stubs.py (hunk 6 above)
Comment (DetachHead) [reply basedtyping/internal_typing_stubs.py]: ive made it private for now and added documentation saying it will be moved

Comment (DetachHead) [reply basedtyping/internal_typing_stubs.py]: moved it to basedmypy's typeshed https://github.com/KotlinIsland/basedmypy/pull/131

File: tests/test_reified_generic.py (hunk 11 above)
Comment (KotlinIsland) [reply tests/test_reified_generic.py]: what about bound TypeVars?

File: basedtyping/runtime_checks.py [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+def issubform(type1: type | UnionType, type2: type | UnionType) -> bool:
+    """``issubclass`` but works with a union as the first arg (like it does for the second arg)
+
+    for example:
+
+    >>> issubclass(int | str, object)
+    TypeError: issubclass() arg 1 must be a class
+    >>> issubtype(int | str, object)
+    True
+    """
```
Comment (KotlinIsland) [inline basedtyping/runtime_checks.py]:
```suggestion
    """Returns ``True`` if ``form`` is a subform (specialform or subclass) of ``forminfo``.
    
    Like  ``issubclass`` but supports typeforms (type-time types)

    >>> issubclass(int | str, object)
    TypeError: issubclass() arg 1 must be a class
    
    >>> issubform(int | str, str)
    False
    
    >>> issubform(int | str, object)
    True
    """
```
Just to make it obvious that it's not matching any entry from the left, it's matching the concept of a union.

Can you update the parameter names as well.

File: basedtyping/runtime_checks.py [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+    if isinstance(type1, UnionType):
```
Comment (KotlinIsland) [inline basedtyping/runtime_checks.py]: What about old union types?

Any any other typeforms maybe `Literal`?

File: basedtyping/reified_generic.py [hunk 15]
```diff
@@ -0,0 +1,160 @@
...
+
+class NotReifiedError(ReifiedGenericError):
+    """raised when a ``ReifiedGeneric`` is instanciated without using a generic alias (ie. the generics can't be reified)"""
+
+
+class UnboundTypeVarError(ReifiedGenericError):
+    """raised when a ``ReifiedGeneric`` is instanciated, ``isinstance`` checked or ``issubclass`` checked with an unbound ``TypeVar``"""
+
+
+class _ReifiedGenericMetaclass(type, OrigClass):
```
Comment (KotlinIsland) [inline basedtyping/reified_generic.py]: I want to have a crack at writing this without a meta class, to be more like the `_GenericAlias` impl

File: basedtyping/reified_generic.py [hunk 16]
```diff
@@ -0,0 +1,160 @@
...
+        try:
+            result.__orig_class__ = self  # type:ignore[attr-defined]
+        except AttributeError:
+            pass
+        return result
+
+    def _check_generics_resolved(self) -> None:
+        if len(self.__parameters__) > 0:
+            raise UnboundTypeVarError(
+                f"generic alias with unbound generics detected: {self.__parameters__}"
```
Comment (KotlinIsland) [inline basedtyping/reified_generic.py]: `unbound` here is wrong I think as the typevar could be bound but not reified.

File: basedtyping/callables.py [hunk 17]
```diff
@@ -0,0 +1,4 @@
+from typing import Callable
+
+Function = Callable[..., object]  # type:ignore[misc]
+"""any ``Callable``. useful when using mypy with ``disallow-any-explicit`` due to https://github.com/python/mypy/issues/9496"""
```
Comment (KotlinIsland) [inline basedtyping/callables.py]: Maybe comment that it's only intended to be used as a bound.

**Review by KotlinIsland — CHANGES_REQUESTED** [review]

File: basedtyping/runtime_checks.py (hunk 14 above)
Comment (DetachHead) [reply basedtyping/runtime_checks.py]: > Any any other typeforms maybe `Literal`?

it currently only works on types and unions, i think it's out of scope for this change to update it to work on `Literal`s and whatever else as well. i tried that once and fell into a rabbithole. i will add a TODO instead

Comment (DetachHead) [reply basedtyping/runtime_checks.py]: > What about old union types?

i added support for them but they're considered `object` by mypy. i've suppressed the error for now because i don't have a clue how to fix it
image

File: basedtyping/runtime_checks.py (hunk 13 above)
Comment (DetachHead) [reply basedtyping/runtime_checks.py]: fixed the function name in the example & added a TODO to make it work on all subforms

File: pyproject.toml [hunk 18]
```diff
@@ -36,5 +37,30 @@ warn_return_any = true
...
+    "fixme",
+    "useless-import-alias",       # required when using no-implicit-reexport in mypy
+    "no-member",                  # handled by mypy
+    "inherit-non-class",          # handled by mypy, false positive on ReifiedGeneric
+    "too-few-public-methods",
+]
+enable = ["useless-suppression", "deprecated-pragma"]
+
+[tool.pylint.FORMAT]
+max-line-length = 200
```
Comment (KotlinIsland) [inline pyproject.toml:72]:
```suggestion
max-line-length = 200

[tool.pytest.ini_options]
addopts = "--import-mode=importlib"

```

**Review by KotlinIsland — CHANGES_REQUESTED** [review]

File: basedtyping/runtime_checks.py (hunk 14 above)
Comment (KotlinIsland) [reply basedtyping/runtime_checks.py]: There's no common type for typeforms and they annotated as plain `object`, maybe it's something to update in typeshed.

File: basedtyping/reified_generic.py [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+            "the generics must be explicitly specified in the instanciation such that it can create a generic alias with the reified generics.\n\n"
+            "for example:\n\n"
+            "foo: Foo[int] = Foo()  # wrong\n"
+            "foo = Foo[int]()  # correct"
+        )
+
+
+class ReifiedGeneric(Generic[T], metaclass=_ReifiedGenericMetaclass):
+    """a ``Generic`` where the types of the ``TypeVars`` are checked to be reified (and visible at type-time),
+    ie. can be accessed at runtime
```
Comment (KotlinIsland) [inline basedtyping/reified_generic.py]:
```
A ``Generic`` where the type parameters are available at runtime and is usable in ``isinstance`` and ``issubclass`` checks.
```

File: basedtyping/reified_generic.py [hunk 20]
```diff
@@ -0,0 +1,161 @@
...
+            *args, **kwargs
+        )
+        try:
+            result.__orig_class__ = self  # type:ignore[attr-defined]
+        except AttributeError:
+            pass
+        return result
+
+    def _check_generics_reified(self) -> None:
+        if len(self.__parameters__) > 0:
```
Comment (KotlinIsland) [inline basedtyping/reified_generic.py]: `if self.__parameters__:`
or 
`if len(self.__parameters__):`

File: basedtyping/reified_generic.py [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+            raise UnboundTypeVarError(
+                f"Type {self.__origin__.__name__} cannot be instantiated; "
+                "generic alias with non-reified generics detected: {self.__parameters__}"
```
Comment (KotlinIsland) [inline basedtyping/reified_generic.py]: Do you think mentioning the internal machinery here is a little confusing?
"Non reified TypeVars cannot be used to instanciate a reified class."

File: basedtyping/reified_generic.py [hunk 22]
```diff
@@ -0,0 +1,160 @@
...
+        ):
+            if not parameter.__contravariant__:
+                # if it's covariant (or invariant in which case check both ways):
+                if not issubform(subclass_arg, self_arg):
+                    return False
+            if not parameter.__covariant__:
+                # if it's contravariant (or invariant in which case check both ways):
+                if not issubform(self_arg, subclass_arg):
+                    return False
+        return True
```
Comment (KotlinIsland) [inline basedtyping/reified_generic.py]: what about something like:
```py
        return all(
            (parameter.__covariant__ and issubform(self_arg, subclass_arg) or True) and
            (parameter.__contravariant__ and issubform(subclass_arg, self_arg) or True)
            for parameter, self_arg, subclass_arg in zip(
                self._type_vars(), self.__args__, args, strict=True
            )
        )
```

All the negations make this hard to understand.

File: basedtyping/reified_generic.py [hunk 23]
```diff
@@ -0,0 +1,160 @@
...
+class ReifiedGenericError(TypeError):
+    pass
+
+
+class NotReifiedError(ReifiedGenericError):
+    """raised when a ``ReifiedGeneric`` is instanciated without using a generic alias (ie. the generics can't be reified)"""
+
+
+class UnboundTypeVarError(ReifiedGenericError):
+    """raised when a ``ReifiedGeneric`` is instanciated, ``isinstance`` checked or ``issubclass`` checked with an unbound ``TypeVar``"""
```
Comment (KotlinIsland) [inline basedtyping/reified_generic.py]:
```suggestion
class NonReifiedTypeVarError(ReifiedGenericError):
    """Raised when a ``ReifiedGeneric`` is instanciated, ``isinstance`` checked or ``issubclass`` checked with a non reified ``TypeVar`` as a type parameter."""
```

Bound `TypeVar`s are still invalid as type parameters to a `ReifiedGeneric`.

**Review by KotlinIsland — CHANGES_REQUESTED** [review]

File: basedtyping/__init__.py [hunk 24]
```diff
@@ -0,0 +1,201 @@
...
+                f"use {self.__origin__.__name__}() instead"
+            )
+        self._check_generics_reified()
+        result = cast(_ReifiedGenericMetaclass, self.__origin__)._actual_call(
+            *args, **kwargs
+        )
+        try:
+            result.__orig_class__ = self  # type:ignore[attr-defined]
+        except AttributeError:
+            pass
```
Comment (KotlinIsland) [inline basedtyping/__init__.py]: `__orig_class__` is basically a necessity right? I think we should fail if we can't set it.

**Review by KotlinIsland — CHANGES_REQUESTED** [review]

File: basedtyping/reified_generic.py (hunk 21 above)
Comment (DetachHead) [reply basedtyping/reified_generic.py]: i think theyre both equally confusing

File: basedtyping/reified_generic.py (hunk 22 above)
Comment (DetachHead) [reply basedtyping/reified_generic.py]: i think thats more difficult to understand, especially with the `or True` parts

File: basedtyping/reified_generic.py (hunk 23 above)
Comment (DetachHead) [reply basedtyping/reified_generic.py]: do u think i should delete one of these errors then? they are for different scenarios but mean the same thing?
https://github.com/KotlinIsland/basedtyping/pull/7/files#diff-7fd05312aa4b91c50fc6766fd2789845da2abc401733491608cca183e50ea6a8R112-R117

Comment (KotlinIsland) [reply basedtyping/reified_generic.py]: One is for `Foo()`, one is for `Foo[T]()`.

If you think that's the same then merge. They do look a little diff to me

Comment (DetachHead) [reply basedtyping/reified_generic.py]: theyre different but you're suggesting the term "not reified" to refer to both scenarios if i understand correctly? i was using "not reified" to refer to `Foo()` and "not bound" to refer to `Foo[T]()` which made sense to me. what do you suggest i rename these two errors to?

File: basedtyping/reified_generic.py (hunk 22 above)
Comment (KotlinIsland) [reply basedtyping/reified_generic.py]: can you think of a less negated way of writing it?
```py
        return all(
            issubform(self_arg, subclass_arg)) if parameter.__covariant__ 
            else issubform(subclass_arg, self_arg) if parameter.__contravariant__
            else subclass_arg == self_arg
            for parameter, self_arg, subclass_arg in zip(
                self._type_vars(), self.__args__, args, strict=True
            )
        )
```

File: basedtyping/reified_generic.py (hunk 21 above)
Comment (KotlinIsland) [reply basedtyping/reified_generic.py]: I don't like it mentioning "generic alias".

File: basedtyping/reified_generic.py [hunk 25]
```diff
@@ -0,0 +1,160 @@
...
+
+
+class _ReifiedGenericMetaclass(type, OrigClass):
+    def _actual_call(cls, *args: NoReturn, **kwargs: NoReturn) -> object:
+        """the actual  ``__call__`` method for the generic alias's ``__origin__``"""
+        return cast(object, super().__call__(*args, **kwargs))
+
+    def __call__(cls, *args: NoReturn, **kwargs: NoReturn) -> object:
+        """a placeholder ``__call__`` method that only gets called if the ``ReifiedGeneric`` being instanciated isn't a ``_ReifiedGenericAlias``"""
+        raise NotReifiedError(
```
Comment (KotlinIsland) [inline basedtyping/reified_generic.py]: So this is only raised when no type parameters are supplied, what about the scenario where 1 / 2 parameters are supplied? Should we check that the correct number are passed?

Comment (DetachHead) [reply basedtyping/reified_generic.py]: i didn't bother adding runtime checks for stuff that mypy would always catch first. do u think i should?

Comment (DetachHead) [reply basedtyping/reified_generic.py]: yea i probably should

Comment (DetachHead) [reply basedtyping/reified_generic.py]: `Generic` already had a check for this so i just added a test to make sure `ReifiedGeneric` still inherits that behavior

File: basedtyping/reified_generic.py (hunk 22 above)
Comment (DetachHead) [reply basedtyping/reified_generic.py]: image

black makes that comprehension completely unreadable + it doesn't work + ratio

File: basedtyping/reified_generic.py (hunk 21 above)
Comment (KotlinIsland) [reply basedtyping/reified_generic.py]: image
also no f on the second line

File: basedtyping/__init__.py [hunk 26]
```diff
@@ -100,70 +105,81 @@ def __instancecheck__(self, instance: object) -> bool:
...
+    """Raised when a ``ReifiedGeneric`` is instantiated without passing type parameters.
 
-    ie. ``foo: Foo[int] = Foo()`` instead of ``foo = Foo[int]()``"""
+    ie: ``foo: Foo[int] = Foo()`` instead of ``foo = Foo[int]()``
+    """
 
 
-class MissingTypeParametersError(ReifiedGenericError):
-    """Raised when a ``ReifiedGeneric`` is instanciated with a non-reified ``TypeVar`` instead of a type parameter
+class NotReifiedError(ReifiedGenericError):
```
Comment (DetachHead) [inline basedtyping/__init__.py]: werent u gonna rename this because `NotReified` also applies to  the scenario covered by `MissingTypeParametersError`

File: basedtyping/__init__.py [hunk 27]
```diff
@@ -100,70 +105,81 @@ def __instancecheck__(self, instance: object) -> bool:
...
         )
 
 
 class ReifiedGeneric(Generic[T], metaclass=_ReifiedGenericMetaclass):
-    """A ``Generic`` where the type parameters are available at runtime and usable in ``isinstance`` and ``issubclass`` checks.
+    """A ``Generic`` where the type parameters are available at runtime and is
+    usable in ``isinstance`` and ``issubclass`` checks.
 
-    for example:
+    Ror example:
```
Comment (DetachHead) [inline basedtyping/__init__.py]: typo

File: basedtyping/__init__.py [hunk 28]
```diff
@@ -183,19 +199,26 @@ def __class_getitem__(cls, item: T) -> type[ReifiedGeneric[T]]:
 
 
 # TODO: make this work with any "form", not just unions
-def issubform(type1: type | UnionType, type2: type | UnionType) -> bool:
-    """``issubclass`` but works with a union as the first arg (like it does for the second arg)
+#  should be (form: TypeForm, forminfo: TypeForm)
+def issubform(form: type | UnionType, forminfo: type | UnionType) -> bool:
+    """Returns ``True`` if ``form`` is a subform (specialform or subclass) of ``forminfo``.
+
+    Like  ``issubclass`` but supports typeforms (type-time types)
```
Comment (DetachHead) [inline basedtyping/__init__.py:221]: i wouldn't add this part until it actually supports more than one typeform

File: basedtyping/__init__.py [hunk 29]
```diff
@@ -183,19 +199,26 @@ def __class_getitem__(cls, item: T) -> type[ReifiedGeneric[T]]:
 
 
 # TODO: make this work with any "form", not just unions
-def issubform(type1: type | UnionType, type2: type | UnionType) -> bool:
-    """``issubclass`` but works with a union as the first arg (like it does for the second arg)
+#  should be (form: TypeForm, forminfo: TypeForm)
+def issubform(form: type | UnionType, forminfo: type | UnionType) -> bool:
```
Comment (DetachHead) [inline basedtyping/__init__.py:216]: why `forminfo`?

File: basedtyping/runtime_only.py [hunk 30]
```diff
@@ -1,11 +1,13 @@
...
 
-this is the equivalent of the `types` module"""
+This is the similar to the ``types`` module.
+"""
 
 from typing import Final, Literal, Union
 
 LiteralType: Final = type(Literal[1])
-"""A type that can be  used to check if type hints are a typing.Literal instance"""
+"""A type that can be  used to check if type hints are a ``typing.Literal``` instance"""
```
Comment (DetachHead) [inline basedtyping/runtime_only.py]: theres an extra `` ` `` after `typing.Literal`

File: tests/test_reified_generics/test_not_reified/test_typevars.py [hunk 31]
```diff
@@ -1,18 +1,18 @@
 """mypy should catch these, but it doesn't due to https://github.com/python/mypy/issues/7084"""
```
Comment (DetachHead) [inline tests/test_reified_generics/test_not_reified/test_typevars.py]: why was this one renamed but `test_bound` wasnt

File: basedtyping/__init__.py (hunk 29 above)
Comment (KotlinIsland) [reply basedtyping/__init__.py:216]: image

File: basedtyping/__init__.py (hunk 28 above)
Comment (KotlinIsland) [reply basedtyping/__init__.py:221]: I added a big warning

File: basedtyping/runtime_only.py [hunk 32]
```diff
@@ -1,11 +1,13 @@
...
 
-this is the equivalent of the `types` module"""
+This is the similar to the ``types`` module.
+"""
 
 from typing import Final, Literal, Union
 
 LiteralType: Final = type(Literal[1])
-"""A type that can be  used to check if type hints are a typing.Literal instance"""
+"""A type that can be  used to check if type hints are a ``typing.Literal`` instance"""
```
Comment (DetachHead) [inline basedtyping/runtime_only.py]: 2 spaces between `be  used`

File: basedtyping/reified_generic.py (hunk 15 above)
Comment (KotlinIsland) [reply basedtyping/reified_generic.py]: Metaclass is needed to always override `__init__`

**Review by KotlinIsland — COMMENTED** [review]
LGTM!

**Review by KotlinIsland — APPROVED** [review]
